train_gpt2.cu

```diff
@@ -390,11 +408,29 @@ void gpt2_write_to_checkpoint(GPT2 *model, const char* checkpoint_path) {
     model_header[6] = model->config.channels;
     model_header[7] = model->config.padded_vocab_size;
     fwriteCheck(model_header, sizeof(int), 256, model_file);
-    // write the parameters
-    device_to_file(model_file, model->params_memory, model->num_parameters_bytes,
-                   IO_BUF_SIZE, main_stream);
-    // close file, we're done
-    fcloseCheck(model_file);
+
+    if (async_write == 0) {
+        // write the parameters
+        device_to_file(model_file, model->params_memory, model->num_parameters_bytes,
+                    IO_BUF_SIZE, main_stream);
+        // close file, we're done
+        fcloseCheck(model_file);
+    }
+    else {
+        // transfer device data to host memory
+        char* buffer_space;
+        cudaCheck(cudaMallocHost(&buffer_space, model->num_parameters_bytes));
```

Comment:
My experience with cudaMallocHost is it can be *much* slower than a regular malloc which eventually "pays for itself" because the data transfer is so much faster, but for a single transfer, I suspect it might be slower than just a regular malloc.

So if that holds true here, this might need to either keep that memory allocated permanently or just use a regular malloc.